include/envoy/http/header_map.h

```diff
@@ -65,9 +71,10 @@ class LowerCaseString {
   void lower() {
     std::transform(string_.begin(), string_.end(), string_.begin(), absl::ascii_tolower);
   }
-  bool valid() const { return validHeaderString(string_); }
+  bool valid() const { return validHeaderString(string_) && validHeaderString(original_string_); }
 
   std::string string_;
+  std::string original_string_;
```

Comment:
I don't think they're gonna accept duplicates of every header name here. They mention in this file they optimize for performance, so having duplicate strings for _every_ header key, even if they didn't opt-in to preserving case, is probably gonna get rejected. Consider only populating this field when the preservation is opted in.